report/topology.go

```diff
@@ -195,6 +197,100 @@ func (c Counters) Copy() Counters {
 	return result
 }
 
+// Sets is a string->set-of-strings map.
+type Sets map[string]StringSet
+
+// Merge merges two sets maps into a fresh set, performing set-union merges as
+// appropriate.
+func (s Sets) Merge(other Sets) Sets {
+	result := s.Copy()
+	for k, v := range other {
+		result[k].Merge(v)
+	}
+	return result
+}
+
+// Copy returns a value copy of the sets map.
+func (s Sets) Copy() Sets {
+	result := Sets{}
+	for k, v := range s {
+		result[k] = v.Copy()
+	}
+	return result
+}
+
+// StringSet is a sorted set of unique strings. Clients must use the Add
+// method to add strings.
+type StringSet []string
```

Comment:
I am not convinced that the ordering provided by `[]string` is a real advantage over using `map[string]struct{}`. 

`map[string]struct{}` would make the implementation much simpler plus ordering can be guaranteed for tests by sorting explicitly.